src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteOutputService.java

```diff
@@ -0,0 +1,623 @@
+    public boolean isDirectory() {
+      return false;
+    }
+
+    @Override
+    public boolean isSymbolicLink() {
+      return true;
+    }
+
+    @Override
+    public boolean isSpecialFile() {
+      return false;
+    }
+  }
+
+  private class GrpcBatchStat implements BatchStat {
+
+    public List<FileStatusWithDigest> batchStat(boolean includeDigest,
+                                                boolean includeLinks,
+                                                Iterable<PathFragment> paths)
+        throws IOException, InterruptedException {
+      // TODO: Do we need to partition the input, just like in
+      // createSymlinkTree(), or is input already guaranteed to be
+      // bounded in size?
+      BatchStatRequest.Builder builder = BatchStatRequest.newBuilder();
+      builder.setBuildId(stateAcrossBuilds.currentBuildId.toString());
+      builder.setIncludeFileDigest(includeDigest);
+      for (PathFragment path : paths) {
+        builder.addPaths(fixupExecRootPath(path));
+      }
+
+      BatchStatResponse responses;
+      try {
+        responses = channel.withChannelBlocking(
+            channel -> RemoteOutputServiceGrpc.newBlockingStub(channel)
+                .batchStat(builder.build()));
+      } catch (StatusRuntimeException e) {
+        throw new IOException(e);
+      }
+      return Lists.newArrayList(
+          Iterables.transform(
+              responses.getResponsesList(),
+              (response) -> {
+                if (!response.hasFileStatus()) {
+                  // File not found.
+                  return null;
+                }
+                FileStatus fileStatus = response.getFileStatus();
+                if (fileStatus.hasFile()) {
+                  Digest digest = fileStatus.getFile().getDigest();
+                  return new RegularFileStatus(digest.getSizeBytes(), DigestUtil.toBinaryDigest(digest));
+                }
+                if (fileStatus.hasDirectory()) {
+                  return new DirectoryFileStatus(fileStatus.getDirectory().getLastModifiedTime().getSeconds());
+                }
+                if (fileStatus.hasSymlink()) {
+                  return new SymlinkFileStatus();
+                }
+                throw new UnsupportedOperationException();
+              }));
+    }
+  }
+
+  @Override
+  public BatchStat getBatchStatter() {
+    return new GrpcBatchStat();
+  }
+
+  @Override
+  public boolean canCreateSymlinkTree() {
+    return true;
+  }
+
+  @Override
+  public void createSymlinkTree(Map<PathFragment, PathFragment> symlinks, PathFragment symlinkTreeRoot)
+      throws ExecException, InterruptedException {
+    // The provided set of symbolic links may be too large to provide to
+    // the remote output service at once. Partition the symbolic links
+    // in groups of 1000, so that BatchCreateRequest messages remain
+    // small enough.
+    UnmodifiableIterator<List<Map.Entry<PathFragment, PathFragment>>> symlinkBatchIterator =
```

Comment:
nit: You can use `Iterables::partition` to spare yourself dealing with the iterator and just use for-each.